site/util.js

```diff
@@ -12,7 +12,77 @@ governing permissions and limitations under the License.
 
 const logger = require('gulplog');
 
-exports.markdown = require('markdown').markdown;
+const md = require('markdown-it')({
+  html: true,
+  linkify: false,
+  typographer: true,
+  breaks: true
+});
+
+function defaultRenderer(tokens, idx, options, env, self) {
+  return self.renderToken(tokens, idx, options, env, self);
+}
+
+let ruleClassnames = {
+  'link_open': 'spectrum-Link',
+  'table_open': 'spectrum-Table',
+  'thead_open': 'spectrum-Table-head',
+  'tr_open': 'spectrum-Table-row',
+  'tbody_open': 'spectrum-Table-body',
+  'th_open': 'spectrum-Table-headCell',
+  'td_open': 'spectrum-Table-cell',
+  'code_inline': 'spectrum-Code spectrum-Code--S',
+  'code_block': 'spectrum-Code spectrum-Code--S'
+};
+
+for (let [rule, className] of Object.entries(ruleClassnames)) {
```

Comment:
yeah this approach requires a lot of processing in util.js.

Reply:
This is applying Spectrum classnames to links, tables, and code elements. It was actually easier than I thought it would be!

Reply:
This is my personal opinion but I always thought having the ids and classnames on the element is more implicit. but that's my personal opinion.

Reply:
I think you misunderstand; this is effectively turning markdown into Spectrum CSS-ified code. We don't want to go and add classes to every instance of `[link](to something else)` or `# Heading` -- instead, we use the renderer architecture of `mardown-it` exactly as it was designed to be used to apply our own custom classes to the rendered HTML markup. We're not writing elements, we're writing markdown, and this is the accepted way to add classes

Reply:
I think my proposed solution is on how to apply those custom classes directly in markdown without the use of a custom processor.

Reply:
So currently, we use markdown in our docs. When we write the following:

```
      Heading is used to create various levels of typographic hierarchies.
      [View the Heading typography component](typography-heading.html)
```

It automatically turns into this:

```
<p>Heading is used to create various levels of typographic hierarchies.</p>
<p><a href="typography-heading.html">View the Heading typography component</a></p>
```

That looks like the following:
![image](https://user-images.githubusercontent.com/201344/74077076-a1d05900-49d1-11ea-9339-ea32d48b07e6.png)

You can see it looks wrong because we're missing the `spectrum-Link` class, right?

So, I've added rules to the markdown parser such that it will output the following HTML:

```
<p>Heading is used to create various levels of typographic hierarchies.</p>
<p><a href="typography-heading.html" class="spectrum-Link">View the Heading typography component</a></p>
```

That looks like the following:

![image](https://user-images.githubusercontent.com/201344/74077095-bdd3fa80-49d1-11ea-8c9e-7eabdcd7952c.png)

Now you can see, it's a real Spectrum link! 

Of course, we could write HTML inside of our markdown and get the same result:

```
      Heading is used to create various levels of typographic hierarchies.
      <a href="typography-heading.html">View the Heading typography component</a>
```

But why would we want to do that? Instead, our markdown is automatically rendered with the correct Spectrum CSS classes.

I did the same for `### Headings`, ``code``, and tables. This is awesome, right?

In the case of the red code blocks, however, I extended the existing code block to add another class if it finds a `~` as the first char. Other approaches with the wrapper were not feasible, and wrapping the tables and using overly specific CSS to target it doesn't make sense.

It's magical and super awesome, and all of the potential solutions either don't work or are tedious and silly. We're using `markdown-it` precisely how it was designed, and there are tons of issues filed against their repository asking how to do exactly what we're doing, documentation that shows how to do exactly what we're doing etc. Again, these aren't utility functions, we're extending the markdown parser's built-in rules and adding our classes to them. There is no performance penalty, and instead we just get beautifully formatted, Spectrum HTML without having to lift a finger. There are zero drawbacks.

Your proposed solution is to simply write HTML within the `yml` files instead of markdown. If that's the case, we might as well completely rip out the markdown parser, edit every instance of a link or backtick code block in every component, and change it all to HTML. We don't want to do that, markdown is here to make writing docs easier, and we should use it.

Reply:
Thanks for the in-depth explanation @lazd . I think I fully understand your approach but I really appreciate your explanation.
In our direct message on Slack I shared with you the example how Boostrap is for example trying to solve this with classnames as well. 
Your approach of applying automatically classnames on HTML elements  might be confusing in the future for someone who is writing our documentation as they might experience classnames being added to elements that they did not expect to be added. This might not be ideal as it might require a previous understanding that post-processing on markdown will be applied. 
For you this might feel like a good solution for simply elements like a header or link. For me it feels like this will get a bit more complex for table and items nested in each other and it may lead to unexpected results and side-effects where wrong styling might need to be applied. 
For example the table classes you are applying are turning the documentation table into an interactive table with hover styles and cursor changes. We would have to apply in utils.js a rule that applies either way an "interactive" class or removes one in the future depending how this might be fixed.

Reply:
All tables are interactive, so even if we wrote all of the HTML hand, it would still have hover styles and cursor changes. See #450, where I have proposed that non-interactive tables are default, and you must add `--interactive` to make them interactive. Once that's implemented, if we ever wanted interactive tables in our examples (which we likely will not), we could simply write the HTML, as you have proposed.

As far as Bootstrap's use of HTML to render tables, we absolutely could do that ourselves, but why? I would expect tables in our markdown to come out looking Spectrum. In addition, Bootstrap doesn't need to add classes to their headings, because Bootstrap indiscriminately  styles naked `<h2>` elements out of the box, and we don't. We specifically require that people opt into our styles by adding classnames, and this change simply makes follow our own recommendation and apply classes to heading, link, and table elements.

The only reason I did not do this in the first place when I added `markdown` to our build process was because it was not possible. Now that we switched to `markdown-it`, this is the obvious, simple approach to generating documentation that matches Spectrum.

That said, I hear you that people might be surprised by Spectrum CSS classes being added automatically. In response to that, I've added a note to the documentation that defines how the metadata is structured. Does that satisfy your concerns?

Reply:
Thank you @lazd  for the clarification again. It sounds like you are proceeding then with adding those classnames to elements automatically and you are suggesting that future users should read more documentation to understand how to document the documentation. 
This still feels a bit complicated to me but that might be my personal feeling about documentation.

Reply:
@bernhard-adobe yes, I believe it's the simplest and most elegant solution. Unless you choose to block the PR, I say we go forward with the Spectrum CSSified markdown and the documentation of how it works.

Reply:
Thanks @lazd , it looks like you are set in your opinion here. I don't think this should be blocking the PR as I feel this discussion might not move forward at this point.